tests/ioctl_lirc.c

```diff
@@ -0,0 +1,167 @@
+
+	errstr = sprintrc(rc);
+
+#ifdef INJECT_RETVAL
+	if (rc != INJECT_RETVAL)
+		error_msg_and_fail("Got a return value of %ld != %ld",
+				   rc, (long) INJECT_RETVAL);
+
+	static char inj_errstr[4096];
+
+	snprintf(inj_errstr, sizeof(inj_errstr), "%s (INJECTED)", errstr);
+	errstr = inj_errstr;
+#endif
+
+	return rc;
+}
+
+int
+main(int argc, char *argv[])
+{
+	int value;
+#ifdef INJECT_RETVAL
+	unsigned long num_skip;
+	bool locked = false;
+
+	if (argc < 2)
+		error_msg_and_fail("Usage: %s NUM_SKIP", argv[0]);
+
+	num_skip = strtoul(argv[1], NULL, 0);
+
+	for (size_t i = 0; i < num_skip; i++) {
+		long ret = ioctl(-1, LIRC_GET_FEATURES, 0);
+
+		printf("ioctl(-1, %s, NULL) = %s%s\n",
+		       XLAT_STR(LIRC_GET_FEATURES), sprintrc(ret),
+		       ret == INJECT_RETVAL ? " (INJECTED)" : "");
+
+		if (ret != INJECT_RETVAL)
+			continue;
+
+		locked = true;
+		break;
+	}
+
+	if (!locked)
+		error_msg_and_fail("Hasn't locked on ioctl(-1"
+				   ", LIRC_GET_FEATURES, NULL) returning %d",
+				   INJECT_RETVAL);
+#endif
+
+	value = 12000;
+	do_ioctl(LIRC_SET_REC_TIMEOUT, &value);
+	printf("ioctl(-1, %s, [12000]) = %s\n",
+	       XLAT_STR(LIRC_SET_REC_TIMEOUT), errstr);
+
+	value = 1;
+	do_ioctl(LIRC_SET_WIDEBAND_RECEIVER, &value);
+	printf("ioctl(-1, %s, [1]) = %s\n",
+	       XLAT_STR(LIRC_SET_WIDEBAND_RECEIVER), errstr);
+
+	value = 0;
+	do_ioctl(LIRC_SET_MEASURE_CARRIER_MODE, &value);
+	printf("ioctl(-1, %s, [0]) = %s\n",
+	       XLAT_STR(LIRC_SET_MEASURE_CARRIER_MODE), errstr);
+
+	value = -1;
+	do_ioctl(LIRC_SET_REC_TIMEOUT_REPORTS, &value);
+	printf("ioctl(-1, %s, [-1]) = %s\n",
+	       XLAT_STR(LIRC_SET_REC_TIMEOUT_REPORTS), errstr);
+
+	value = 33;
+	do_ioctl(LIRC_SET_SEND_DUTY_CYCLE, &value);
+	printf("ioctl(-1, %s, [33]) = %s\n",
+	       XLAT_STR(LIRC_SET_SEND_DUTY_CYCLE), errstr);
+
+	value = 38000;
+	do_ioctl(LIRC_SET_SEND_CARRIER, &value);
+	printf("ioctl(-1, %s, [38000]) = %s\n",
+	       XLAT_STR(LIRC_SET_SEND_CARRIER), errstr);
+
+	value = 20000;
+	do_ioctl(LIRC_SET_REC_CARRIER, &value);
+	printf("ioctl(-1, %s, [20000]) = %s\n",
+	       XLAT_STR(LIRC_SET_REC_CARRIER), errstr);
+
+	value = 40000;
+	do_ioctl(LIRC_SET_REC_CARRIER_RANGE, &value);
+	printf("ioctl(-1, IPMICTL_SET_MAINTENANCE_MODE_CMD or %s, [40000]) = %s\n",
+	       XLAT_STR(LIRC_SET_REC_CARRIER_RANGE), errstr);
+
+	value = 2;
+	do_ioctl(LIRC_SET_SEND_MODE, &value);
+	printf("ioctl(-1, %s, [LIRC_MODE_PULSE]) = %s\n",
+	       XLAT_STR(LIRC_SET_SEND_MODE), errstr);
+
+	value = 8;
+	do_ioctl(LIRC_SET_SEND_MODE, &value);
+	printf("ioctl(-1, %s, [LIRC_MODE_SCANCODE]) = %s\n",
+	       XLAT_STR(LIRC_SET_SEND_MODE), errstr);
+
+	value = 4;
+	do_ioctl(LIRC_SET_REC_MODE, &value);
+	printf("ioctl(-1, %s, [LIRC_MODE_MODE2]) = %s\n",
+	       XLAT_STR(LIRC_SET_REC_MODE), errstr);
+
+	value = 16;
+	do_ioctl(LIRC_SET_REC_MODE, &value);
+	printf("ioctl(-1, %s, [LIRC_MODE_LIRCCODE]) = %s\n",
+	       XLAT_STR(LIRC_SET_REC_MODE), errstr);
+
+	// read ioctl
```

Comment:
/* read ioctls */

To make the test work in case of no syscall tampering,
the return value of do_ioctl should be tested and acted appropriately,
see, for eexample, tests/ioctl_gpio.c

also, please add `do_ioctl(LIRC_GET_FEATURES, 0);` or something like this to cover the `true` branch of `if (umove_or_printaddr(tcp, arg, &value))`.

Reply:
OK, there is `ioctl(-1, LIRC_GET_FEATURES, 0)` in `INJECT_RETVAL` variant, it's probably enough.
At least the coverage test reports that all lines are covered.

Reply:
I've added this case. Please let me know what you think.